File: app/src/main/java/workshop/akbolatss/tagsnews/base/BaseActivity.java
package workshop.akbolatss.tagsnews.base;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import workshop.akbolatss.tagsnews.application.App;
import workshop.akbolatss.tagsnews.di.component.AppComponent;

public abstract class BaseActivity extends AppCompatActivity {

    private Unbinder unbinder;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getContentView());
        unbinder = ButterKnife.bind(this);

        onViewReady(savedInstanceState, getIntent());
    }

    @CallSuper
    protected void onViewReady(Bundle savedInstanceState, Intent intent) {
        //For Child activities
    }

    @Override
    protected void onDestroy() {
        unbinder.unbind();
        super.onDestroy();
    }

    protected AppComponent getAppComponent() {
        return ((App) getApplication()).getAppComponent();
    }

    protected abstract int getContentView();
}

File: app/src/main/java/workshop/akbolatss/tagsnews/base/BaseActivity.kt
package workshop.akbolatss.tagsnews.base

import android.content.Intent
import android.os.Bundle
import android.support.annotation.CallSuper
import me.yokeyword.swipebackfragment.SwipeBackActivity
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.application.App
import workshop.akbolatss.tagsnews.di.component.AppComponent
import workshop.akbolatss.tagsnews.screen.splash.SplashActivity

/**
 * Base class for Activities in project. Extends by #SwipeBackActivity to handle Swipe to close gesture
 */
abstract class BaseActivity : SwipeBackActivity() {

    protected val appComponent: AppComponent
        get() = (application as App).appComponent

    protected abstract fun getContentView(): Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getContentView())

        //Exclusion for SplashActivity
        if (this !is SplashActivity)
            overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_right)

        onInitDagger()
        onViewReady(savedInstanceState, intent)
    }

    /**
     * Same transition animation for child Activities
     */
    override fun onBackPressed() {
        super.onBackPressed()
        overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_right)
    }

    /**
     * Instantiate Dagger2
     */
    protected open fun onInitDagger() {}


    @CallSuper
    protected open fun onViewReady(savedInstanceState: Bundle?, intent: Intent) {
        //For Child activities
    }
}
